KAFKA-4617: Generate core project with correct source folders.

dhwanikatagade opened the pull request:
Generate core project with correct source folders. In addition set output folders same as command line build. Don't generate unnecessary projects (https://issues.apache.org/jira/browse/KAFKA-4604)

Commit:
KAFKA-4617: Generate core project with correct source folders. In addition set output folders same as command line build. Don't generate unnecessary projects.

dhwanikatagade commented:
@ijuma @guozhangwang - I have opened this PR that obsoletes https://github.com/apache/kafka/pull/2324
The tests run on my setup with transient failures which pass on re-running.

edoardocomar commented:
Hi @dhwanikatagade thanks for the patch. 
The source folders in core are now correct.
I see that you have made the eclipse compiled classes location match the ones used by gradle on the CLI. 

However two generated Eclipse projects (`core` and `streams`) still have a mistake.

Their generated classpaths still include as _library_ entries the output directories of other projects they depend on. they should not

e.g. for `core` these library entries must not be there 
```
	<classpathentry kind="lib" path="/Users/ecomar/devel/ws-scratch/kafka/clients/build/classes/test"/>
	<classpathentry kind="lib" path="/Users/ecomar/devel/ws-scratch/kafka/clients/build/resources/test"/>
``` 
because the classpath already (correctly) contains this 

```
<classpathentry kind="src" path="/clients"/>
```

and similarly for the `streams` project these must not be present:
```
	<classpathentry kind="lib" path="/Users/ecomar/devel/ws-scratch/kafka/clients/build/classes/test"/>
	<classpathentry kind="lib" path="/Users/ecomar/devel/ws-scratch/kafka/clients/build/resources/test"/>
	<classpathentry kind="lib" path="/Users/ecomar/devel/ws-scratch/kafka/core/build/classes/test"/>
	<classpathentry kind="lib" path="/Users/ecomar/devel/ws-scratch/kafka/core/build/resources/test"/>
```

because the `streams` project already depends also on `core` and `clients` and it correctly has 
```
	<classpathentry kind="src" path="/clients"/>
	<classpathentry kind="src" path="/json"/>
	<classpathentry kind="src" path="/api"/>
	<classpathentry kind="src" path="/core"/>

```

Commit:
Merge branch 'trunk' into gradle_eclipse_plugin_path_fix

Commit:
KAFKA-4617: Generate core project with correct source folders. In addition set output folders same as command line build. Don't generate unnecessary projects and don't add test output directories of dependencies projects.

dhwanikatagade commented:
@edoardocomar thanks for your comments. 
I tried to remove the following explicit dependency from build.gradle but it is required by the scala plugin for building test classes. Without it the :core:compileTestScala task fails.
```
testCompile project(':clients').sourceSets.test.output
``` 
I had to fix the generated .classpath file in the whenMerged closure itself. I have pushed an additional commit with changes to handle this.
Please review.

edoardocomar commented:
@dhwanikatagade thanks for the fix
I tested the patch on a fresh workspace. All Eclipse projects created nicely.
LGTM 
@ijuma I would love to have this fix in

dhwanikatagade commented:
Hi @guozhangwang  @ijuma could you please merge this PR. 
The tests pass on my dev setup.
If for some reasons we would not want to have the PR merged please let me know and I will close the PR.

dhwanikatagade commented:
retest this please

Commit:
Merge branch 'trunk' into gradle_eclipse_plugin_path_fix

guozhangwang commented:
retest this please

guozhangwang commented:
Looks good to me. cc @ijuma

ijuma commented:
In the original gradle thread, a number of issues were raised if the same output directory is used by Eclipse and Gradle:

https://discuss.gradle.org/t/eclipse-generated-files-should-be-put-in-the-same-place-as-the-gradle-generated-files/6986/2

It may be worth adding a gradle property to allow people to choose the behaviour. Regarding the default, it would seem that the safe option should be the default (`bin` is clearly not a good default though).

ijuma commented on build.gradle:
```diff
@@ -117,6 +116,33 @@ if (new File('.git').exists()) {
+  if (!project.name.equals('connect')) {
+    apply plugin: 'eclipse'
+
+    eclipse.classpath.file {
```
It's a good idea to add a comment explaining the reasoning for this. Also, maybe it should be in a separate method.

Commit:
Merge branch 'trunk' into gradle_eclipse_plugin_path_fix

dhwanikatagade commented:
Thanks @ijuma 
The reasons why I had over looked the mentioned problems with having a common build output directory between gradle and eclipse were
* A clean build resolved the problems in Eclipse and well as Gradle. I do it often in Eclipse when things look messed up.
* Keeping common output folders helped me get the Eclipse dev setup working.
* I have always had common output folders between maven and eclipse.

All the same I will look into the possibility getting my setup to work with the following changes from your review suggestions
1. Having a separate output folder for eclipse (say build_eclipse)
2. Refactoring the in-line instructions

And of course comments :-)

Commit:
Merge branch 'trunk' into gradle_eclipse_plugin_path_fix

Commit:
Generate core project with correct source folders. In addition set eclipse output folders to 'build_eclipse'.  Don't generate unnecessary projects and don't add test output directories of dependencies projects.

dhwanikatagade commented:
Hi @ijuma 
I have made the following changes
1. Set eclipse output to go to build_eclipse directory
2. Moved the classpath pruning code to a method
3. Added comments explaining purpose of the changes

Please review.

dhwanikatagade commented:
retest this please

Commit:
Merge branch 'trunk' into gradle_eclipse_plugin_path_fix

dhwanikatagade commented:
@ijuma @guozhangwang - Could you please review this once and see if it is worthwhile to merge.

Commit:
Merge branch 'trunk' into gradle_eclipse_plugin_path_fix

Commit:
Generate core project with correct source folders. In addition set eclipse output folders to 'build_eclipse'.  Don't generate unnecessary projects and don't add test output directories of dependencies projects.

guozhangwang commented:
LGTM. cc @ijuma

ijuma commented:
@edoardocomar, can you verify that this works for you?

@dhwanikatagade, is it worth adding a note to the "Building IDE" section in the readme?

edoardocomar commented:
Hi @ijuma seen your request, I had just retested it from scratch. 

The generated eclipse projects work way better than what we had, huge improvements. Even the core project has the nested sourcepaths correct to run unit tests.

the only issue I get is a compilation error in streams
```
Description	Resource	Path	Location	Type
Type mismatch: cannot convert from KeyValue<capture#27-of ? extends K1,V> to KeyValue<K1,V>	KStreamImpl.java	/streams/src/main/java/org/apache/kafka/streams/kstream/internals	line 145	Java Problem
```
curious what that is

dhwanikatagade commented:
> @dhwanikatagade, is it worth adding a note to the "Building IDE" section in the readme?

@ijuma - Once this change is applied I only have to run the `./gradlew eclipse` command. Would you want me to add notes to the README explaining that the build output goes to build_eclipse and why?

> the only issue I get is a compilation error in streams
> Description	Resource	Path	Location	Type
> Type mismatch: cannot convert from KeyValue<capture#27-of ? extends K1,V> to KeyValue<K1,V>	KStreamImpl.java	/streams/src/main/java/org/apache/kafka/streams/kstream/internals	line 145	Java Problem
> curious what that is

@edoardocomar - This is due to an Eclipse issue that is observed even without the the changes in place. I just have to add a cast as a local change as follows to get it working. 
```
-                        return new KeyValue<>(mapper.apply(key, value), value);
+                        return (KeyValue<K1, V>) new KeyValue<>(mapper.apply(key, value), value);
```
I have known this issue in Eclipse where the maven build would work, IntelliJ would not report any errors, yet Eclipse would complain.

ijuma commented:
@dhwanikatagade, yeah, I was wondering if it was worth adding a short note stating that the output directory is `build_eclipse` in order to avoid clashes with gradle and our bin scripts.

dhwanikatagade commented:
> @dhwanikatagade, yeah, I was wondering if it was worth adding a short note stating that the output directory is build_eclipse in order to avoid clashes with gradle and our bin scripts.

@ijuma - I will provide that shortly in a push.

Commit:
Merge branch 'trunk' into gradle_eclipse_plugin_path_fix

Commit:
Generate core project with correct source folders. In addition set eclipse output folders to 'build_eclipse'.  Don't generate unnecessary projects and don't add test output directories of dependencies projects. Added a corresponding note to README.

ijuma reviewed (approved):
Thanks, LGTM. Merging to trunk after minor edits to the readme.

dhwanikatagade commented:
> Merging to trunk after minor edits to the readme.
@ijuma - Thanks for the edits. It's neater now. :-)
